cpp/include/raft/util/cuda_utils.cuh

```diff
@@ -508,44 +508,170 @@ HDI double myATanh(double x)
 /** @} */
 
 /**
- * @defgroup LambdaOps Lambda operations in reduction kernels
+ * @defgroup LambdaOps Commonly used lambda operations
  * @{
  */
-// IdxType mostly to be used for MainLambda in *Reduction kernels
+// The optional index argument is mostly to be used for MainLambda in reduction kernels
 template <typename Type, typename IdxType = int>
 struct Nop {
-  HDI Type operator()(Type in, IdxType i = 0) { return in; }
+  HDI Type operator()(Type in, IdxType i = 0) const { return in; }
+};
+
+struct KeyOp {
+  template <typename KVP, typename IdxType = int>
+  HDI typename KVP::Key operator()(const KVP& p, IdxType i = 0) const
+  {
+    return p.key;
+  }
+};
+
+struct ValueOp {
+  template <typename KVP, typename IdxType = int>
+  HDI typename KVP::Value operator()(const KVP& p, IdxType i = 0) const
+  {
+    return p.value;
+  }
 };
 
 template <typename Type, typename IdxType = int>
 struct SqrtOp {
-  HDI Type operator()(Type in, IdxType i = 0) { return mySqrt(in); }
+  HDI Type operator()(Type in, IdxType i = 0) const { return mySqrt(in); }
 };
 
 template <typename Type, typename IdxType = int>
 struct L0Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return in != Type(0) ? Type(1) : Type(0); }
+  HDI Type operator()(Type in, IdxType i = 0) const { return in != Type(0) ? Type(1) : Type(0); }
 };
 
 template <typename Type, typename IdxType = int>
 struct L1Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return myAbs(in); }
+  HDI Type operator()(Type in, IdxType i = 0) const { return myAbs(in); }
 };
 
 template <typename Type, typename IdxType = int>
 struct L2Op {
-  HDI Type operator()(Type in, IdxType i = 0) { return in * in; }
+  HDI Type operator()(Type in, IdxType i = 0) const { return in * in; }
 };
 
-template <typename Type>
+template <typename InT, typename OutT = InT>
 struct Sum {
-  HDI Type operator()(Type a, Type b) { return a + b; }
+  HDI OutT operator()(InT a, InT b) const { return a + b; }
+};
+
+template <typename InT, typename OutT = InT>
+struct Subtract {
```

Comment:
Since the effort is to unify the use of these structs and encourage their use, I'd suggest we revisit https://github.com/rapidsai/raft/issues/1024 again and move the template arguments into the operator() where possible. I'd leave at the struct level only in non-inferable cases, like the non-type template parameters or return type.


Although I see it can be tempting to do the arithmetic op and convert data type in one go, I think it's not healthy to have the `OutT` different from the `InT` in the arithmetic functions like these. Instead, we could provide more "functors" like the [`ann_utils/mapping` ](https://github.com/rapidsai/raft/blob/branch-22.12/cpp/include/raft/spatial/knn/detail/ann_utils.cuh#L158) plus the function composition (something like this):
```
template <typename OuterF, typename InnerF>
struct compose {
  OuterF fo;
  InnerF fi;
  template <typename... Args>
  HDI auto operator()(Args... args) const { return fo(fi(args...); }
};
```
It could be worth checking if the resulting compiled code is efficient though...

Reply:
I agree, hence my comment above. Mainly we need to figure out the proper way of deprecating template args for the functors that already existed in the API. Regarding the input/output types, I mostly agree it's not really useful to have different input/output types if it only does an implicit cast anyway. For non-trivial conversions, we could add a composition operator like your `compose` structure. For trivial casts, I think it's unnecessary.

Reply:
That's one possibility, yes. And we can also use the occasion to improve the naming scheme (`Sum` -> `add_op`, `L2Op` -> `square_op`, `L1Op` -> `abs_op` etc ; it's all open for discussion)

@cjnolet What do you think about this? (deprecating the old functors and replacing them with new type-inferred ones with different names).

Reply:
@achirkin I've added automatic type inference and deprecated the existing functors.